feat!: Enable pprof server on demand (#16594) (#17068)
argoproj/argo-cd: docs/operator-manual/pprof.md

> # Profiling with pprof

Reviewer: ```suggestion
# Profiling with `pprof`
```

> pprof is a powerful tool for profiling Go applications. It is part of the Go standard library and provides rich insights into the performance characteristics of Go programs. Profiling with pprof involves gathering runtime statistics about the application's execution, such as CPU usage, memory allocation, and contention.

Reviewer: ```suggestion
[`pprof`](https://go.dev/blog/pprof) is a powerful tool for profiling Go applications. It is part of the Go standard library and provides rich insights into the performance characteristics of Go programs, such as CPU usage, memory allocation, and contention.
```
- link to `pprof` blog / usage description here
- use backticks for commands etc, per [k8s style guide](https://kubernetes.io/docs/contribute/style/style-guide/#use-code-style-for-kubernetes-command-tool-and-component-names)
- simpler and more direct, per [k8s style guide](https://kubernetes.io/docs/contribute/style/style-guide/#use-simple-and-direct-language)

> ## Installation
> 
> pprof is included in the Go standard library and integrated in all ArgoCD components, so no separate installation is required. To enable the profiling endpoints, you just need to set the environment variable `ARGO_PPROF` with the port that you want to use for exposing the endpoints. e.g: `ARGO_PPROF=8888` will start the profiling endpoints using the port `8888`.
> 
> ## Basic Usage

Reviewer: ```suggestion
## Basic Usage

Enable profiling endpoints by setting the environment variable `ARGO_PPROF` with your preferred port. For instance, `ARGO_PPROF=8888` will start profiling endpoints on port `8888`.

```

- remove "Installation" entirely as it is redundant / not needed. the env var is part of the usage
  - no need to repeat "Go standard library" in this section either
- simpler and more direct, per [k8s style guide](https://kubernetes.io/docs/contribute/style/style-guide/#use-code-style-for-kubernetes-command-tool-and-component-names)
- avoid words like "just", per [k8s style guide](https://kubernetes.io/docs/contribute/style/style-guide/#avoid-words-that-assume-a-specific-level-of-understanding)
- avoid Latin phrases, per [k8s style guide](https://kubernetes.io/docs/contribute/style/style-guide/#avoid-latin-phrases)
- use active voice, per [k8s style guide](https://kubernetes.io/docs/contribute/style/style-guide/#use-active-voice)

- add a new line after the heading, per `markdownlint`

> pprof operates in two modes: interactive and non-interactive. In the non-interactive mode, it generates profiling data that can be analyzed later. In the interactive mode, it launches a web server to visualize the profiling data in real-time.

Reviewer: ```suggestion
`pprof` has two modes: interactive and non-interactive. Non-interactive mode generates profiling data for future analysis. Interactive mode launches a web server to visualize the profiling data in real-time.
```

- use backticks for commands, per k8s style guide
- simpler and more direct, per k8s style guide
- use active voice, per k8s style guide

> > Note: Replace http://localhost:6060 with the appropriate URL of your Go application.

Reviewer: ```suggestion
!! Note "Replace URLs"
    In the below commands, replace http://localhost:6060 with the appropriate URL of your Argo component.
```
- use `mkdocs-material` [admonition](https://squidfunk.github.io/mkdocs-material/reference/admonitions/)
- it wasn't clear that you were referring to the below commands, so specify that
- "your Go Application" -> "your Argo component"

> ### Generating CPU Profiles
> To generate a CPU profile for a Go application, you can use the following command:

Reviewer: ```suggestion
### Generate CPU Profile

Generate a CPU profile with the following command:
```

- use active voice, per k8s style guide
- remove "for a Go application"; not needed
- both of these make it simpler and more direct, per k8s style guide

- also newline after heading, per `markdownlint`

```bash
go tool pprof http://localhost:6060/debug/pprof/profile
```

> ### Generating Heap Profiles
> To generate a heap profile, you can use the following command:

Reviewer: ```suggestion
### Generate Heap Profiles

Generate a heap profile with the following command:
```

same as above comment

```bash
go tool pprof http://localhost:6060/debug/pprof/heap
```

> ### Interactive Mode
> To run pprof in interactive mode, you can use the following command:

Reviewer: ```suggestion
### Interactive Mode

Use interactive mode with the following command:
```

- same as above comment
- this is all about `pprof` already, so remove redundant reference.
  - and modes are already described above, so assume the reader knows what you mean

```bash
go tool pprof -http=:8080 http://localhost:6060/debug/pprof/profile
```

> This command launches a web server and opens a browser window displaying the profiling data.

Reviewer: ```suggestion

This starts a web server and opens a browser window displaying the profiling data.
```

- newline before & after all code blocks, per `markdownlint`
- "starts" is more common wording than "launches" (see also "avoid jargon" in [k8s style guide](https://kubernetes.io/docs/contribute/style/style-guide/#avoid-jargon-and-idioms))
- simplify a bit

> ## More info
> 
> To extend the information about how to profile golang applications, you could find interesting [this article](https://go.dev/blog/pprof) from golang blog.

Reviewer: ```suggestion
```

linked above now, so no more need for this section, which also makes it simpler and more direct.
